### PR TITLE
 Provide capability to use UTC time notation for console logs

### PR DESCRIPTION
**Pull request by tomkerkhove**
Provide capability to use UTC time notation for console logs via `ConsoleLoggerOptions.UseUtcTimezone`.

This is turned off by default to be backwards compatible.

Addresses #1808

In terms of testing I would suggest to provide a similar test to `WriteCore_LogsCorrectTimestamp` which uses UTC option and verifies stamp or a similar one - Thoughts?

### CODE REVIEW

File: src/Logging/Logging.Console/src/ConsoleLoggerOptions.cs [hunk 1]
```diff
@@ -37,5 +37,10 @@ public ConsoleLoggerFormat Format
         /// Gets or sets format string used to format timestamp in logging messages. Defaults to <c>null</c>.
         /// </summary>
         public string TimestampFormat { get; set; }
+
+        /// <summary>
+        /// Gets or sets indication whether or not UTC timezone should be used to for timestamps in logging messages. Defaults to <c>false</c>.
+        /// </summary>
+        public bool UseUtcTimezone { get; set; } = false;
```
Comment (gfoidl) [inline src/Logging/Logging.Console/src/ConsoleLoggerOptions.cs]:
```suggestion
        public bool UseUtcTimezone { get; set; }
```
`false` is already the default.

File: src/Logging/Logging.Console/src/ConsoleLogger.cs [hunk 2]
```diff
@@ -137,11 +137,16 @@ private LogMessageEntry CreateDefaultLogMessage(StringBuilder logBuilder, LogLev
                 logBuilder.AppendLine(exception.ToString());
             }
 
+            string timestamp = null;
             var timestampFormat = Options.TimestampFormat;
+            if (timestampFormat != null)
+            {
+                timestamp = Options.UseUtcTimezone ? DateTime.UtcNow.ToString(timestampFormat) : DateTime.Now.ToString(timestampFormat);
```
Comment (gfoidl) [inline src/Logging/Logging.Console/src/ConsoleLogger.cs]:
```suggestion
                var dateTime = Options.UseUtcTimezone ? DateTime.UtcNow : DateTime.Now;
                timestamp = dateTime.ToString(timestampFormat);
```
But I think this is more a matter of taste.

File: src/Logging/Logging.Console/src/ConsoleLoggerOptions.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Logging/Logging.Console/src/ConsoleLoggerOptions.cs]: True but I wanted to be more explicit - However it doesn't really make sense so I'll remove it!

File: src/Logging/Logging.Console/src/ConsoleLogger.cs (hunk 2 above)
Comment (tomkerkhove) [reply src/Logging/Logging.Console/src/ConsoleLogger.cs]: Fine for me, will send update asap

**Comment by tomkerkhove** [comment]
@gfoidl I've incorporated your suggested and added a test similar to the existing one.

However, they do not verify the output so I'm curious how you would approach it?
Should I just deserialize them and verify the timezone? Open to suggestions.

File: src/Logging/test/ConsoleLoggerTest.cs [hunk 3]
```diff
@@ -382,14 +382,50 @@ public void WriteCore_LogsCorrectTimestamp(ConsoleLoggerFormat format, LogLevel
...
+            var t = SetUp(new ConsoleLoggerOptions { TimestampFormat = "yyyyMMddHHmmss ", Format = format, UseUtcTimezone = true});
+            var levelPrefix = t.GetLevelPrefix(level);
+            var logger = t.Logger;
+            var sink = t.Sink;
+            var ex = new Exception("Exception message" + Environment.NewLine + "with a second line");
+
+            // Act
+            logger.Log(level, 0, _state, ex, _defaultFormatter);
+
+            // Assert
```
Comment (gfoidl) [inline src/Logging/test/ConsoleLoggerTest.cs]: Hm, there is no assert that UTC is written, so this test is more or less a duplication of the already existing test for normal datetime. (just see you commented on that too :wink:).

> Should I just deserialize them and verify the timezone?

That's possible (but then you have to assert / take into account that current timezone != UTC).

Maybe it's better to (at least in my projects I would do) extract the creation of the `timestamp` in `ConsoleLogger` to a separate method that is testable?

Something like 
```c#
private static string GetTimestamp(string timestampFormat, bool useUtcTimezone)
    => GetTimestamp(timestampFormat, useUtcTimezone, DateTime.Now);

internal static string GetTimestamp(string timestampFormat, bool useUtcTimezone, DateTime now)
{
    if (timestampFormat == null)
        return null;

    if (useUtcTimezone)
        now = now.ToUniversalTime();

    return now.ToString(timestampFormat);
}
```
Then the second method is testable.

Comment (tomkerkhove) [reply src/Logging/test/ConsoleLoggerTest.cs]: That's one approach but who says that method is being called? We have no guarantee about that actually so not sure if that's going to cover everything.

Comment (gfoidl) [reply src/Logging/test/ConsoleLoggerTest.cs]: You're right. So your approach is a better end-to-end test.

Comment (tomkerkhove) [reply src/Logging/test/ConsoleLoggerTest.cs]: So what I'd propose then would be:
1. Change format to `yyyyMMddHHmmsszzz`
2. Get tracked timestamp via regex similar to https://github.com/aspnet/Extensions/pull/1896/files#diff-a27bf784d2879ac5f61a0d1a8573fc80R418
3. Parse to DateTime and check timezone

Thoughts?

I'm not a regex expert so help would be ideal on that 😅

Comment (tomkerkhove) [reply src/Logging/test/ConsoleLoggerTest.cs]: This should be there, except for explicitly setting the culture to a non-UTC compatible one for the non-UTC test.

I tried to use the folling but it didn't work:
```csharp
var cultureInfo = new CultureInfo("en-US");
CultureInfo.DefaultThreadCurrentCulture = cultureInfo;
CultureInfo.DefaultThreadCurrentUICulture = cultureInfo;
```

**Comment by tomkerkhove** [comment]
Looks like my PR is messed up due to the rebase, going to start with a clean branch.

**Comment by tomkerkhove** [comment]
Closing in favor of #2420 with a clean branch.

**Pull request closed without merging**